File: 416-partition-equal-subset-sum/partition-equal-subset-sum.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
// Bottom-Up Dynamic Programming
// Time and Space: O(n * m)....n for # of nums and m for half the sum of nums
var canPartition = function(nums) {
    let sum = nums.reduce((acc, v) => acc + v), n = nums.length;
    if (sum & 1) return false;
    sum = Math.floor(sum / 2);

    let dp = [...new Array(n+1)].map(a => 
             [...new Array(sum+1)].fill(false))
    for (let arr of dp) arr[0] = true;

    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= sum; j++) {

            if (j - nums[i - 1] >= 0) {
                dp[i][j] = dp[i-1][j - nums[i - 1]]
            }
            dp[i][j] = dp[i-1][j] || dp[i][j]
        }
    }
    return dp[n][sum]
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int n = nums.length; int sum = 0;
        for (int num : nums) sum += num;
        
        if (sum % 2 != 0) return false;
        sum = (int) Math.floor(sum / 2);
        boolean[][] dp = new boolean[n+1][sum+1];
        for (boolean[] arr: dp) arr[0] = true;

        for (int r = 1; r < dp.length; r++) {
            for (int c = 1; c < dp[r].length; c++) {
                if (c - nums[r-1] >= 0) {
                    dp[r][c] = dp[r-1][c - nums[r-1]] || dp[r-1][c];
                } else {
                    dp[r][c] = dp[r-1][c];
                }
            }
        }
        return dp[dp.length-1][dp[0].length-1];
    }
}